### ISSUE TITLE
exec-env/exec-file: support "--" to separate command to run

### ISSUE THREAD

**Issue by ancoron** [getsops/sops#1469]
An issue I am facing regularly is that I need to create an actual command that is being executed with `exec-env` or `exec-file` with variables. Due to the need to put the whole command line into a single argument it becomes very and error-prone due to quoting issues.

A simple example:
```bash
my_url="${api_baseuri}/${resource_path}"
sops exec-env test.env 'curl -s --user "${API_USER}:${API_PASSWORD}" "'"${my_url}"'"'
```

If sops could introduce special support for the `--` (double-dash) argument, we could say that everything after it becomes a part of the command to execute and it could look like this instead:
```bash
sops exec-env test.env -- curl -s --user '${API_USER}:${API_PASSWORD}' "${my_url}"
```

For manual/interactive command building, this would also allow to auto-complete file names in the shell we're working in.

Of course, the user would still need to ensure that variable expansion doesn't happen in the shell (using single quotes instead of double-quotes), but that has less error potential than quoting (at least for me).

**Comment by felixfontein** [comment]
Hmm, this is not so trivial to implement. Simply passing on the array of strings from the command line arguments on as a new command is simple, but that doesn't allow you to access further shell features (like `${API_USER}`). But if you want to pass what has been handed in through another shell, we need to figure out how to handle other things that shell might interpret, such as `!` or spcaes in arguments. For example, what should `sops exec-env test.env -- foo "bar baz" "b'a'm"` execute? The command `foo` with three parameters `bar`, `baz`, and `b'a'm`? Or the command `foo` with two arguments `bar baz` and `b'a'm`? And do you want to quote the single quotes in `b'a'm`? Or remove them as a shell would do?

**Comment by ancoron** [comment]
@felixfontein Thanx for the feedback and I agree, that behavior needs to be well-defined to clarify expectations on the user side.

Picking up your example:
```bash
sops exec-env test.env -- foo "bar baz" "b'a'm"
```

...should internally be creating a command with the following parts:
1. command/executable: `foo`
2. command arguments:
   * ` bar baz`
   * `b'a'm`
 
To sops itself, the argument should present itself as follows:
1. `exec-env`
2. `test.env`
3. `--`
4. `foo`
5. `bar baz`
6. `b'a'm`

So its barely a pass-through of arguments with the exception of the first argument after `--` to represent the command/executable.

If I had a shell script where I needed to produce an actual command line, I'd have something like this:
```bash
my_url="${api_baseuri}/${resource_path}"

curl_opts=(-s --fail-with-body --user '${API_USER}:${API_PASSWORD}')

# ...add other options depending on script behavior...

sops exec-env test.env -- curl "${curl_opts[@]}" "${my_url}"
```

Here, the `"${curl_opts[@]}"` is of course expanded by the shell into the individual arguments as presented in the array and due to single-quoting the user option value, it will expand neither `API_USER` nor `API_PASSWORD`, which then shall be handled by sops.

At least that's my take on this issue.

**Comment by uasi** [comment]
@ancoron @felixfontein Why not just make sops execute the given arguments, and use an actual shell if we need to expand variables, like `sops exec-env test.env -- bash -c 'curl -s --user "${API_USER}:${API_PASSWORD}" "{my_url}"'` ?

**Comment by ancoron** [comment]
@uasi The main point of this issue is to support `--` in the first place with a well-defined argument handling that doesn't get in the way.

In the description I've presented briefly what the main issue with SOPS is when it comes to dynamic command building as part of scripting.

In your example I'd still need to:
1. export variable `my_url` so the `sops` sees it
2. trust that `sops exec-env` will actually pass it down to the command
3. somehow assume that `test.env` does **NOT** contain variable `my_url`

The major issue is with variable expansion and control of **WHERE** it gets expanded without losing readability of the SOPS invocation code.

**Comment by joshbode** [comment]
My quick hack to work around not being able to pass multiple arguments to `sops exec-env` (until it is supported) is this wrapper function:

```shell
_sops_exec() {
  local FILE=$1 COMMAND=$2 ARGS
  shift 2

  ARGS=$(printf " %q" "$@")
  sops exec-env "${FILE}" "${COMMAND}${ARGS}"
}
alias sops-exec=_sops_exec
```

which can be used like:

```shell
$ sops-exec ~/secrets.env foo --bar --baz x y z
```

**Comment by robinanterior** [comment]
Proposed fix in #1626 

PS @joshbode fyi bash has that functionality built-in using the `@Q` modified:

```
$ foo="Joe's   tavern"
$ echo "${foo@Q}"
'Joe'\''s   tavern'
$ echo ${foo@Q}
'Joe'\''s tavern'
$ args "${foo@Q}"
["'Joe'\\''s   tavern'"]
$ args ${foo@Q}
["'Joe'\\''s", "tavern'"]
```
```

**Comment by hraban** [comment]
Proposed fix in #1626

PS @joshbode fyi bash has that functionality built-in using the @Q modifier:

```
$ foo="Joe's   tavern"
$ echo "${foo@Q}"
'Joe'\''s   tavern'
$ echo ${foo@Q}
'Joe'\''s tavern'
$ args "${foo@Q}"
["'Joe'\\''s   tavern'"]
$ args ${foo@Q}
["'Joe'\\''s", "tavern'"]
```

**Comment by peter-leonov-ch** [comment]
Would be nice to have a simple version of this, much like most of other wrapper tools do: `npm run tsc -- --noEmit`, `kubectl exec -it pod/xxx -- bash -l -c "env"`, `yarn shell -- -t` and more.

**Comment by jscheid** [comment]
For reasons described in my [other comment](https://github.com/getsops/sops/issues/1368#issuecomment-3789969301) this should be doing the same thing as all similar tools (e.g. sudo), which is simply to pass the arguments to `execv[pe]`. 

In short, the reasons are:
- rule of least surprise (works like other similar tools)
- simple mental model (no hidden machinery)
- fewest foot guns (minimizes opportunity for accidental misuse)

All of which are of consequence for a security tool.

So, `sops exec-env secrets.env cat foo` simply becomes `execvpe("cat", ["cat", "foo"], new_env)`.

That enables a safe, robust, one-line solution for the original ask:

```sh
sops exec-env test.env sh -c 'exec curl --silent --user="${API_USER}:${API_PASSWORD}" "$0"' "$my_url"
```

That's safe, unlike the original example:

```sh
my_url='http://example.com" ; echo $API_PASSWORD > /tmp/PWNED ; #'
env API_PASSWORD=s3kr1t sh -c 'curl -s --user "${API_USER}:${API_PASSWORD}" "'"${my_url}"'"' # don't do this
cat /tmp/PWNED
```

Personally I would probably create a helper script instead of the one-liner, say `curl-with-auth`, for clarity:

```sh
#/bin/sh
exec /usr/bin/curl --user="${API_USER}:${API_PASSWORD}" "$@"
```

```sh
sops exec-env test.env ./curl-with-auth --silent "$my_url"
```

**Comment by ashb** [comment]
> So, `sops exec-env secrets.env` cat foo simply becomes `execvpe("cat", ["cat", "foo"], new_env)`.

Agreed with this, and very much  "rule of least suprise" -- I didn't read the docs and assumed it would work this way, and was very confused when I ended up with errors of "error: missing file to decrypt" when I tried it.

**Comment by KingMichaelPark** [comment]
I have just taken to doing:
```zsh
ai() {
    local args="$*"
    sops exec-env "$HOME/.dotfiles/access.age.json" "opencode $args"
}
```
